### PR TITLE
Make Targeting Opt-in

### LINKED ISSUE #12307: [Improvement]: Targeting should use opt-in approach
### Improvement description

Currently, the [Targeting and Personalization](https://pimcore.com/docs/pimcore/current/Development_Documentation/Tools_and_Features/Targeting_and_Personalization/index.html#page_Opt-out-from-targeting) feature is enabled "by default". That is, an opt-_out_ approach is used as soon as `isTargetingConfigured` returns `true`:

https://github.com/pimcore/pimcore/blob/c73a4641ba6b815f1e5787e642591284b982a92c/lib/Targeting/EventListener/TargetingListener.php#L134-L136

> The opt-out approach is also documented [here](https://pimcore.com/docs/pimcore/current/Development_Documentation/Tools_and_Features/Targeting_and_Personalization/index.html#page_Opt-out-from-targeting).

Instead of this, an opt-_in_ approach should be used. For example, instead of `pimcore_targeting_disabled=1`, introduce something like a `pimcore_targeting_enabled=1` cookie and enable the targeting only if this cookie was set.

Using tools like Cookiebot, the user can now choose to consent to the _marketing_-cookies (which are needed according to Cookiebot). If the user consents, the `pimcore_targeting_enabled=1` cookie can be set, which will then enable the actual targeting functionality (starting with the next request).

### PR DESCRIPTION
**Pull request by aryaantony92**
## Changes in this pull request  
Resolves #12307

### CODE REVIEW

File: bundles/PersonalizationBundle/src/Targeting/EventListener/TargetingEventListener.php [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+    use EnabledTrait;
+
+    public static function getSubscribedEvents(): array
+    {
+        return [
+            KernelEvents::REQUEST => ['onKernelRequest']
+        ];
+    }
+
+    public function onKernelRequest(RequestEvent $event): void
```
Comment (martineiber) [inline bundles/PersonalizationBundle/src/Targeting/EventListener/TargetingEventListener.php]: What's the reason we moved this into a separate file?
Now we have two almost identical files? `TargetingEventListener.php` and `TargetingListener.php`
I think we have a better overview if we merge this into one file? wdyt?

File: bundles/PersonalizationBundle/config/targeting/services.yaml [hunk 2]
```diff
@@ -4,39 +4,8 @@ services:
...
-        arguments:
-            $bundle: "@=service('kernel').getBundle('PimcorePersonalizationBundle')"
 
-    # auto-register all controllers as services
-    Pimcore\Bundle\PersonalizationBundle\Controller\:
-        resource: '../src/Controller'
-        public: true
-        tags: [ 'controller.service_arguments' ]
 
     Pimcore\Bundle\PersonalizationBundle\Targeting\Maintenance\TargetingStorageTask: ~
```
Comment (martineiber) [inline bundles/PersonalizationBundle/config/targeting/services.yaml:9]: This should also be moved into the targeting.yaml

File: bundles/PersonalizationBundle/config/targeting.yaml [hunk 3]
```diff
@@ -4,6 +4,20 @@ services:
         autoconfigure: true
         public: false
 
+    #
+    # INSTALLER
+    #
+    Pimcore\Bundle\PersonalizationBundle\Installer:
```
Comment (martineiber) [inline bundles/PersonalizationBundle/config/targeting.yaml]: That has nothing to do with targeting, right?
This is something for the PersonalizationBundle itself? So it would be better if we move this to a service.yaml/installer.yaml. Same for the controllers and Newsletter Adapter. 

If we have a targeting.yaml, it should only contain services explicitly for trargeting.

File: bundles/PersonalizationBundle/src/DependencyInjection/Configuration.php [hunk 4]
```diff
@@ -36,9 +36,10 @@ public function getConfigTreeBuilder(): TreeBuilder
         $rootNode
             ->children()
                 ->arrayNode('targeting')
-                    ->canBeDisabled()
-                    ->addDefaultsIfNotSet()
```
Comment (martineiber) [inline bundles/PersonalizationBundle/src/DependencyInjection/Configuration.php:40]: Why did you remove `addDefaultsIfNotSet`?

File: bundles/PersonalizationBundle/src/Targeting/EventListener/TargetingEventListener.php (hunk 1 above)
Comment (aryaantony92) [reply bundles/PersonalizationBundle/src/Targeting/EventListener/TargetingEventListener.php]: It's moved to a diffetent name because the TargetingListener.php would be loaded only if the personalization bundle is enabled. To avoid the confusion, the file is renamed to TargetingEnableListener in [041b52d](https://github.com/pimcore/pimcore/pull/14462/commits/041b52d81d5e8199b35a579ac2f31ec90e74cdc8)

File: bundles/PersonalizationBundle/config/targeting.yaml (hunk 3 above)
Comment (aryaantony92) [reply bundles/PersonalizationBundle/config/targeting.yaml]: Moved to config/services.yaml

**Comment by martineiber** [comment]
It make no different if `pimcore_targeting_enabled` cookie is set or not. 
Could you please check this again.

**Comment by aryaantony92** [comment]
We have kicked the cookie approach. So, targeting is disabled by default now and is enabled, only when it's enabled from the config

File: bundles/PersonalizationBundle/config/targeting.yaml [hunk 5]
```diff
@@ -160,3 +160,32 @@ services:
...
+    #
+
+    Pimcore\Bundle\PersonalizationBundle\EventListener\IndexSettingsListener:
+        tags:
+            - { name: kernel.event_listener, event: pimcore.admin.indexAction.settings, method: indexSettings }
+
+    Pimcore\Bundle\PersonalizationBundle\Targeting\EventListener\Frontend\TargetingElementListener: ~
+    Pimcore\Bundle\PersonalizationBundle\Targeting\EventListener\Frontend\TargetingDocumentRendererListener: ~
+    Pimcore\Bundle\PersonalizationBundle\Targeting\EventListener\RenderletListener: ~
+    Pimcore\Bundle\PersonalizationBundle\Targeting\EventListener\TargetingEnableListener: ~
```
Comment (martineiber) [inline bundles/PersonalizationBundle/config/targeting.yaml]: This line can be removed. Since the Listener was also removed. 
```suggestion
```

**Comment by martineiber** [comment]
Should we also disable this button if `pimcore_personalization:targeting:enabled:` is set to false? 
It is a bit strange when the button is visible but the functionality is not active. 

image

**Comment by fashxp** [comment]
Hmm but there needs to be an opt-option for the User!
Globally enabling via config is not enough!
It is about gdpr and cookie guidelines

**Review by fashxp — CHANGES_REQUESTED** [review]
See my comment above

**Comment by fashxp** [comment]
Also Docs need to be updated accordingly! 
We need to explain somewhere what needs to be done that targeting geht's started

**Comment by martineiber** [comment]
Changing the value of `pimcore_targeting_enabled` has no effect on the toolbar. 
If the toolbar is enabled in the backend (`pimcore_targeting_debug=1` cookie is set), it is not possible to disable the toolbar with `pimcore_targeting_enabled=0`.

The same behavior is when the toolbar is disabled in the backend, it is not possible to enable the toolbar with the `pimcore_targeting_enabled=1` cookie set.

**Comment by aryaantony92** [comment]
> Changing the value of `pimcore_targeting_enabled` has no effect on the toolbar. If the toolbar is enabled in the backend (`pimcore_targeting_debug=1` cookie is set), it is not possible to disable the toolbar with `pimcore_targeting_enabled=0`.
> 
> The same behavior is when the toolbar is disabled in the backend, it is not possible to enable the toolbar with the `pimcore_targeting_enabled=1` cookie set.

Changed the check in [4857b86](https://github.com/pimcore/pimcore/pull/14462/commits/4857b86fb1ac4aace09f738f57695e799f8debce). So, now toolbar will be disabled with pimcore_targeting_enabled=0 if it's enabled in the backend.

**Comment by dvesh3** [comment]
> > Changing the value of `pimcore_targeting_enabled` has no effect on the toolbar. If the toolbar is enabled in the backend (`pimcore_targeting_debug=1` cookie is set), it is not possible to disable the toolbar with `pimcore_targeting_enabled=0`.
> > The same behavior is when the toolbar is disabled in the backend, it is not possible to enable the toolbar with the `pimcore_targeting_enabled=1` cookie set.
> 
> Changed the check in [4857b86](https://github.com/pimcore/pimcore/pull/14462/commits/4857b86fb1ac4aace09f738f57695e799f8debce). So, now toolbar will be disabled with pimcore_targeting_enabled=0 if it's enabled in the backend.

@aryaantony92 This is a bug and should be fixed on 10.5

File: bundles/PersonalizationBundle/doc/Development_Documentation/README.md [hunk 6]
```diff
@@ -156,7 +156,10 @@ bar):
 
 * <a href="javascript:(function()%7Bdocument.cookie %3D 'pimcore_targeting_debug%3D1%3B path%3D%2F'%7D)()">Enable Pimcore Targeting Toolbar</a> 
 
-## Opt-out from targeting
-You can give the user the possibility to opt-out from targeting at any time, by setting the following cookie: `pimcore_targeting_disabled=1`. 
+## Opt-in for Targeting
+
+<div class="alert alert-warning">
+Targeting is now using the opt-in approach and will not be enabled by default. So, user should set the following cookie `pimcore_targeting_enabled=1` to enable targeting. They can use tools like `Cookiebot` to choose consent and set the cookie accordingly.
```
Comment (dvesh3) [inline bundles/PersonalizationBundle/doc/Development_Documentation/README.md]:
```suggestion
A user has the possibility to enable targeting at any time, by setting the following cookie `pimcore_targeting_enabled=1` .
```
It should not sound like upgrade note in the docs.

File: bundles/PersonalizationBundle/src/Targeting/EventListener/TargetingListener.php [hunk 7]
```diff
@@ -85,15 +85,15 @@ public static function getSubscribedEvents(): array
...
     public function onKernelRequest(RequestEvent $event): void
     {
-        if (!$this->enabled) {
-            return;
-        }
-
-        if ($event->getRequest()->cookies->has('pimcore_targeting_disabled')) {
-            $this->disable();
+        if($event->getRequest()->cookies->has('pimcore_targeting_enabled')) {
+            if($event->getRequest()->cookies->get('pimcore_targeting_enabled')) {
```
Comment (dvesh3) [inline bundles/PersonalizationBundle/src/Targeting/EventListener/TargetingListener.php]:
```suggestion
        if($event->getRequest()->cookies->getBoolean('pimcore_targeting_enabled')) {
```

File: bundles/PersonalizationBundle/src/Targeting/EventListener/Traits/EnabledTrait.php [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+    }
+
+    public function isEnabled(): bool
+    {
+        return $this->enabled;
+    }
+
+    protected function isTargetingEnabled(): bool
+    {
+        if(\Pimcore::getKernel()->getContainer()->getParameter('pimcore_personalization.targeting.enabled') && $this->enabled) {
```
Comment (dvesh3) [inline bundles/PersonalizationBundle/src/Targeting/EventListener/Traits/EnabledTrait.php]: this returns false if the cookie sets the `enabled` property and config doesn't...right?

Comment (aryaantony92) [reply bundles/PersonalizationBundle/src/Targeting/EventListener/Traits/EnabledTrait.php]: Yes

File: bundles/PersonalizationBundle/src/Targeting/Service/TargetingEnableService.php [hunk 9]
```diff
@@ -0,0 +1,47 @@
...
+
+    public function isTargetingEnabled(): bool
+    {
+        return $this->enabled;
+    }
+
+    public function enableTargeting(): bool {
+        $request = $this->requestHelper->getCurrentRequest();
+
+        if($request->cookies->getBoolean('pimcore_targeting_enabled') || \Pimcore::getKernel()->getContainer()->getParameter('pimcore_personalization.targeting.enabled')) {
```
Comment (dvesh3) [inline bundles/PersonalizationBundle/src/Targeting/Service/TargetingEnableService.php]: I would set the enable state from the `PimcorePersonalizationExtension::configureTargeting` based on the config `enabled` instead of checking parameter `pimcore_personalization.targeting.enabled`.

Comment (dvesh3) [reply bundles/PersonalizationBundle/src/Targeting/Service/TargetingEnableService.php]: and return true if `$this->enabled || $request->cookies->getBoolean('pimcore_targeting_enabled')`

File: bundles/PersonalizationBundle/src/Targeting/EventListener/TargetingSessionBagListener.php [hunk 10]
```diff
@@ -54,7 +58,11 @@ public static function getSubscribedEvents(): array
 
     public function onKernelRequest(RequestEvent $event): void
     {
-        if (!$this->isEnabled()) {
+        if(!$this->targetingEnableService->enableTargeting()) {
+            return;
+        }
+
+        if(!$this->isEnabled()) {
```
Comment (dvesh3) [inline bundles/PersonalizationBundle/src/Targeting/EventListener/TargetingSessionBagListener.php:65]: Is this check still required? since enabled is checked from the service.

File: bundles/PersonalizationBundle/src/Targeting/Service/TargetingEnableService.php [hunk 11]
```diff
@@ -0,0 +1,47 @@
...
+    {
+        $this->requestHelper = $requestHelper;
+    }
+
+    public function isTargetingEnabled(): bool
+    {
+        return $this->enabled;
+    }
+
+    public function enableTargeting(): bool {
```
Comment (dvesh3) [inline bundles/PersonalizationBundle/src/Targeting/Service/TargetingEnableService.php]:
```suggestion
    public function isTargetingEnabled(): bool {
```
get rid of the other method

File: doc/23_Installation_and_Upgrade/09_Upgrade_Notes/README.md [hunk 12]
```diff
@@ -270,6 +268,7 @@ pimcore:
 - Moved implementation of `PimcoreBundleAdminSupportInterface` from `AbstractPimcoreBundle` to bundle classes.
     Moved `getJsPaths`, `getCssPaths`, `getEditmodeJsPaths` and `getEditmodeCssPaths` from `AbstractPimcoreBundle` to `BundleAdminSupportTrait`.
 - [Cache] Responses containing a header `Cache-Control: no-cache`, `Cache-Control: private` or `Cache-Control: no-store` will no longer be cached by the full page cache.
+- [Targeting] Targeting is now using the opt-in approach and will not be enabled by default. So, user should set the following cookie `pimcore_targeting_enabled=1` to enable targeting. User can use tools like `Cookiebot` to choose consent and set the cookie accordingly.
```
Comment (dvesh3) [inline doc/23_Installation_and_Upgrade/09_Upgrade_Notes/README.md]: Move this under `- [Personalization and targeting]` and add sample config to enable targeting.

File: bundles/PersonalizationBundle/src/Targeting/EventListener/TargetingSessionBagListener.php (hunk 10 above)
Comment (aryaantony92) [reply bundles/PersonalizationBundle/src/Targeting/EventListener/TargetingSessionBagListener.php:65]: This check is checking the session.enabled https://github.com/pimcore/pimcore/blob/b81da81fb5514e813a4d446a855477eebf86f9a3/bundles/PersonalizationBundle/src/Targeting/EventListener/TargetingSessionBagListener.php#L152

File: doc/23_Installation_and_Upgrade/09_Upgrade_Notes/README.md [hunk 13]
```diff
@@ -150,6 +150,14 @@
...
     - [Personalization and targeting] has been moved to `pimcore/personalization` package.
       - Config `pimcore:targeting:` has been removed, please use `pimcore_personalization.targeting` in the PimcorePersonalizationBundle instead.
+      - Targeting is now using the opt-in approach and will not be enabled by default. So, user should set the following cookie `pimcore_targeting_enabled=1` to enable targeting. User can use tools like`Cookiebot` to choose consent and set the cookie accordingly.
+       Targeting can also be enabled by setting the following config.
+      ```yaml
+      pimcore_personalization:
+          targeting:
+              enabled: true
+      ```
+
```
Comment (dvesh3) [inline doc/23_Installation_and_Upgrade/09_Upgrade_Notes/README.md]:
```suggestion
      - Targeting is now using the opt-in approach and will not be enabled by default. Add following config to enable it:
      ```yaml
      pimcore_personalization:
          targeting:
              enabled: true
      ```
      Also, a user can use cookie `pimcore_targeting_enabled=1` to enable targeting.
```

**Review by dvesh3 — CHANGES_REQUESTED** [review]
Adapt Docs to enable targeting as it does not work out of the box anymore https://github.com/pimcore/pimcore/blob/2809776dab7c1ad6955ac85c429534ce1b8c489c/bundles/PersonalizationBundle/doc/Development_Documentation/README.md#setup

File: doc/23_Installation_and_Upgrade/09_Upgrade_Notes/README.md [hunk 14]
```diff
@@ -257,8 +264,6 @@ pimcore:
         valid_languages:
             - en
             - de
-```
```
Comment (dvesh3) [inline doc/23_Installation_and_Upgrade/09_Upgrade_Notes/README.md:260]: please revert this change

**Review by dvesh3 — APPROVED** [review]
